contracts/eosio.system/src/eosio.system.cpp

```diff
@@ -500,6 +588,67 @@ namespace eosiosystem {
       return false;
    }
 
+   // -------------------------------------------------------------------------------------------
+   struct canon_name_t
+   {
+      uint64_t _value; // all significant characters are shifted to the least significant position
+      uint64_t _size;  // number of significant characters
+
+      uint64_t mask() const { return (1ull << (_size * 5)) - 1; }
+      uint64_t size() const { return _size; }
+
+      // starts from the end, so n[0] is the last significant character of the name
+      uint64_t operator[](uint64_t i) const { return (_value >> (5 * i)) & 0x1F; }
+
+      canon_name_t(name n) {
+         uint64_t v         = n.value;
+         auto     num_zeros = std::countr_zero(v);
+         check(num_zeros >= 4, "Account names and patterns should be less than 13 characters long");
+         
+         uint64_t shift = 4 + ((num_zeros - 4) / 5) * 5;
+         _value  = v >> shift;
+         auto sig_bits = 64 - shift;
+         assert(sig_bits % 5 == 0);
+         _size = sig_bits / 5;
+         assert((_value & ~mask()) == 0);
+         assert((_value & mask()) == _value);
+      }
+
+      // returns true if account_name ends with the same characters as `this`, preceded by a `.` (zero) character
+      bool is_suffix_of(canon_name_t account_name) const {
+         assert(_size < account_name._size);
+         if (((_value ^ account_name._value) & mask()) != 0)
+            return false;
+
+         // pattern matches the end of account_name. To be a suffix it has to be preceded by a dot in account_name
+         return account_name[size()] == 0;
+      }
+
+      bool found_in(canon_name_t account_name) const {
+         assert(_size < account_name._size);
+         auto   sz   = size();
+         size_t diff = account_name.size() - sz + 1;
+         for (size_t i = 0; i < diff; ++i)
+            if (((_value ^ (account_name._value >> (5 * i))) & mask()) == 0)
+               return true;
+         return false;
+      }
+   };
+
+   // -------------------------------------------------------------------------------------------
+   bool name_allowed(name account, name patt) {
+      canon_name_t pattern(patt);
+      canon_name_t account_name(account);
+      
+      if (pattern.size() >= account_name.size())
+         return true;
+      if (pattern.is_suffix_of(account_name))
+         return true;
+      if (pattern.found_in(account_name))
+         return false;
+      return true;
+   }
```

Comment:
Would be nice to have some unittests of just the `cannon_name_t`. It is difficult to reason that it is correct.

Reply:
I'd have to duplicate the code though, as `canon_name_t` is defined in `eosio.system.cpp`, and I don't think we want to move it to `eosio.system.hpp`.

Reply:
The tests would not pass if `canon_name_t` was wrong. 
Here is a compiler explorer version (`canon_name_t` is named `pattern_t` there) https://godbolt.org/z/3Mqo397vP

Reply:
Could be moved to a `cannon_name.hpp` so available for testing. It does seem like additional test cases would be good regardless.